src/python/pants/testutil/rule_runner.py

```diff
@@ -656,89 +659,163 @@ class MockGet(Generic[_O]):
+        (expected_calls if e.rule_id else expected_gets).append(e)
+
+    expected_call_names = {str(e.rule_id) for e in expected_calls}
+    actual_call_names = set(actual_calls.keys())
+
+    missing_call_names = expected_call_names - actual_call_names
+    surplus_call_names = actual_call_names - expected_call_names
 
     def as_comparable(
-        o: Get | Effect | MockGet | MockEffect | AwaitableConstraints | type,
+        o: MockGet | MockEffect | AwaitableConstraints,
     ) -> MockRequestExceptionComparable:
-        if isinstance(o, MockGet) or isinstance(o, Get):
+        category: Literal["Get", "Effect"]
+
+        if isinstance(o, MockGet):
             category = "Get"
-        elif isinstance(o, MockEffect) or isinstance(o, Effect):
+        elif isinstance(o, MockEffect):
             category = "Effect"
         elif isinstance(o, AwaitableConstraints):
             category = "Effect" if o.is_effect else "Get"
         else:
-            # If we don't know of this class, try to give something meaningful
-            # This happened with AwaitableContraints
-            maybe_output_type = getattr(o, "output_type", None)
-            maybe_input_types = getattr(o, "input_types", None)
-            return MockRequestExceptionComparable(
-                category=f"Uncategorised of type {type(o).__qualname__}",
-                output_type=maybe_output_type.__name__ if maybe_output_type is not None else None,
-                input_types=(
-                    tuple(e.__name__ for e in maybe_input_types)
-                    if maybe_input_types is not None
-                    else tuple()
-                ),
+            raise ValueError(
+                f"Unsupported object {o} of type {type(o)} passed where a "
+                "MockGet, MockEffect or AwaitableConstraints was expected"
             )
 
-        output_type = o.output_type.__name__
-        input_types = tuple(e.__name__ for e in o.input_types)
-
         return MockRequestExceptionComparable(
             category=category,
-            output_type=output_type,
-            input_types=input_types,
+            output_type=o.output_type,
+            input_types=o.input_types,
         )
 
-    expected_as_comparable = {as_comparable(e) for e in expected}
-    actual_as_comparable = {as_comparable(e) for e in actual}
+    expected_gets_as_comparable = {as_comparable(e) for e in expected_gets}
+    actual_gets_as_comparable = {as_comparable(e) for e in actual_gets}
 
-    missing = expected_as_comparable - actual_as_comparable
-    additional = actual_as_comparable - expected_as_comparable
+    missing_gets = expected_gets_as_comparable - actual_gets_as_comparable
+    surplus_gets = actual_gets_as_comparable - expected_gets_as_comparable
 
-    return (
-        f"HINT: missing : {[str(e) for e in missing]}\n"
-        f"HINT: additional : {[str(e) for e in additional]}\n"
-    )
+    if union_membership:
+        for mg in list(missing_gets):
```

Comment:
I feel like this could use a comment - it's a hard read

Reply:
Yeah, I will add one. The fact that the previous code got it backwards is a testament to this being hard to grok.